sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,62 +27,270 @@ cdef inline double fmax(double x, double y) nogil:
     return y
 
 
+cdef double _logsumexp(double* arr, int n_classes) nogil:
+    """Computes the sum of arr assuming arr is in the log domain.
+
+    Returns log(sum(exp(arr))) while minimizing the possibility of
+    over/underflow.
+    """
+    # Use the max to normalize, as with the log this is what accumulates
+    # the less errors
+    cdef double vmax = arr[0]
+    cdef double out = 0.0
+    cdef int i
+
+    for i in range(1, n_classes):
+        if vmax < arr[i]:
+            vmax = arr[i]
+
+    for i in range(n_classes):
+        out += exp(arr[i] - vmax)
+
+    return log(out) + vmax
+
+
+cdef class MultinomialLogLoss:
+    cdef double _loss(self, double* prediction, double y, int n_classes,
+                      double sample_weight) nogil:
+        """Multinomial Logistic regression loss.
+
+        The multinomial logistic loss for one sample is:
+        loss = - sw \sum_c \delta_{y,c} (prediction[c] - logsumexp(prediction))
+             = sw (logsumexp(prediction) - prediction[y])
+
+        where:
+            prediction = dot(x_sample, weights) + intercept
+            \delta_{y,c} = 1 if (y == c) else 0
+            sw = sample_weight
+
+        Parameters
+        ----------
+        prediction : pointer to a np.ndarray[double] of shape (n_classes,)
+            Prediction of the multinomial classifier, for current sample.
+
+        y : double, between 0 and n_classes - 1
+            Indice of the correct class for current sample.
+
+        n_classes : integer
+            Total number of classes.
+
+        sample_weight : double
+            Weight of current sample.
+
+        Returns
+        -------
+        loss : double
+            Multinomial loss for current sample.
+
+        Reference
+        ---------
+        Bishop, C. M. (2006). Pattern recognition and machine learning.
+        Springer. (Chapter 4.3.4)
+        """
+        cdef double logsumexp_prediction = _logsumexp(prediction, n_classes)
+        cdef double loss
+
+        # y is the indice of the correct class of current sample.
+        loss = (logsumexp_prediction - prediction[int(y)]) * sample_weight
+        return loss
+
+    cdef void _dloss(self, double* prediction, double y, int n_classes,
+                     double sample_weight, double* gradient_ptr) nogil:
+        """Multinomial Logistic regression gradient of the loss.
+
+        The gradient of the multinomial logistic loss for one sample is:
```

Comment:
`loss` is the loss, `dloss` is the derivative of the loss, has used in sgd_fast.pyx